crates/evm/src/lib.rs

```diff
@@ -121,3 +124,123 @@ pub trait ConfigureEvmEnv: Send + Sync + Unpin + Clone + 'static {
         fill_block_env(block_env, chain_spec, header, after_merge);
     }
 }
+
+/// Trait for managing the EVM context.
+/// Allows to build an EVM Executor with the given context.
+pub trait EvmContext<'a, EXT, DB: Database> {
+    /// The executor produced with the context set.
+    type Executor: EvmExecutor;
```

Comment:
I'm not sure if we actually need this

Reply:
We had thought of this to have a building feature that just does the job of building an EVM instance with env, a database and an inspector. This was mainly intended to be used in situations calling for `evm_config.evm_with_env(&mut db, env)` like here for example:

https://github.com/paradigmxyz/reth/blob/579da56e9792a45b52b4a77929b43d3da995c9d9/crates/ethereum/payload/src/lib.rs#L348

Or

https://github.com/paradigmxyz/reth/blob/579da56e9792a45b52b4a77929b43d3da995c9d9/crates/rpc/rpc/src/eth/api/transactions.rs#L561

I've also thought about the following formulation with decoupled traits to be able to implement:
- either the trait with `with_db`
- either the trait with `with_inspector`
- either the trait with all the env context
- or all 3 at the same time.

```rust
pub trait EvmDatabaseContext<DBN: Database> {
 typeDatabaseOutput;

 fn with_db(self, db: DBN) -> Self::DatabaseOutput;
}

pub trait EvmInspectorContext<EXTN> {
 typeInspectorOutput;

 fn with_inspector(self, inspector: EXTN) -> Self::InspectorOutput;
}

pub trait EvmContext {
 /// The associated type for the executor.
 Executor type: EvmExecutor;

 /// Sets the EVM environment.
 ///
 /// This method returns a new EVM context with the given environment.
 #[must_use]
 fn with_env(self, env: Env) -> Self;

 /// Sets the EVM configuration environment.
 ///
 /// This method returns a new EVM context with the given configuration environment.
 #[must_use]
 fn with_cfg_env(self, cfg: CfgEnv) -> Self;

 /// Sets the block environment.
 ///
 /// This method returns a new EVM context with the given block environment.
 #[must_use]
 fn with_block_env(self, block: BlockEnv) -> Self;

 /// Sets the transaction environment.
 ///
 /// This method returns a new EVM context with the given transaction environment.
 #[must_use]
 fn with_tx_env(self, tx: TxEnv) -> Self;

 /// Sets the specification (hardfork) for the EVM instance.
 ///
 /// This method returns a new EVM context with the given specification (hardfork).
 #[must_use]
 fn with_spec_id(self, spec_id: SpecId) -> Self;

 /// Returns the EVM executor.
 ///
 /// This method returns the executor configured with the current context.
 fn executor(self) -> Self::Executor;
}

pub trait UnifiedEvmContext<DBN: Database, EXTN>:
 EvmDatabaseContext<DBN> + EvmInspectorContext<EXTN> + EvmContext
{
}

impl<'a, EXT, DB: Database, DBN: Database + 'a, EXTN: 'a> UnifiedEvmContext<DBN, EXTN>
 for RevmContext<'a, EXT, DB>
{
}

impl<'a, EXT, DB: Database, DBN: Database + 'a> EvmDatabaseContext<DBN>
 for RevmContext<'a, EXT, DB>
{
 type DatabaseOutput = RevmContext<'a, EXT, DBN>;

 fn with_db(self, db: DBN) -> Self::DatabaseOutput {
 let context = revm::Context::new(self.context.evm.with_db(db), self.context.external);
 let handler = revm::Handler::new(self.handler.cfg());
 RevmContext { context, handler }
 }
}

impl<'a, EXT, EXTN: 'a, DB: Database> EvmInspectorContext<EXTN> for RevmContext<'a, EXT, DB> {
 type InspectorOutput = RevmContext<'a, EXTN, DB>;

 fn with_inspector(self, inspector: EXTN) -> Self::InspectorOutput {
 let context = revm::Context::new(self.context.evm, inspector);
 let handler = revm::Handler::new(self.handler.cfg());
 RevmContext { context, handler }
 }
}

impl<'a, EXT, DB: Database> EvmContext for RevmContext<'a, EXT, DB> {
 Executor type = Evm<'a, EXT, DB>;

 fn with_env(mut self, env: Env) -> Self {
 *self.context.evm.env() = env;
 self
 }

 fn with_cfg_env(mut self, cfg: CfgEnv) -> Self {
 self.context.evm.env().cfg = cfg;
 self
 }

 fn with_block_env(mut self, block: BlockEnv) -> Self {
 self.context.evm.env().block = block;
 self
 }

 fn with_tx_env(mut self, tx: TxEnv) -> Self {
 self.context.evm.env().tx = tx;
 self
 }

 fn with_spec_id(mut self, spec_id: SpecId) -> Self {
 self.handler.modify_spec_id(spec_id);
 self
 }

 fn executor(self) -> Self::Executor {
 Evm { context: self.context, handler: self.handler }
 }
}
```

What do you think of this approach? I understand that with the `BlockExecutorProvider` we are able to return an executor but I don't see how we can generically replace all the places where we need the builder?

Reply:
> replace all the places where we need the builder?

do we even need the builder? rn we also never return it from configureevm.

from rpc pov we only need something that can transact (multiple times)

Reply:
If you believe this can be done without a builder, let's keep the `EvmTransact` and `EvmCommit` traits, and then we can check in another PR how easy it is to update the RPC